components/esp_lvgl_port/src/lvgl9/ppa/bsp_dma2d.c

```diff
@@ -0,0 +1,170 @@
+#include "freertos/task.h"
+#include "dma2d_utils.h"
+#include "esp_cache.h"
+#include "esp_timer.h"
+#include "bsp_dma2d.h"
+
+static char *TAG = "convert";
+
+#define ALIGN_UP(num, align)    (((num) + ((align) - 1)) & ~((align) - 1))
+#define ALIGN_DOWN(num, align)    ((num) & (~((align) - 1)))
+
+static size_t get_cache_line_size(const void *addr)
+{
+    esp_err_t ret = ESP_FAIL;
+    size_t cache_line_size = 0;
+    uint32_t heap_caps = esp_ptr_external_ram(addr) ? MALLOC_CAP_SPIRAM : MALLOC_CAP_INTERNAL;
+    ret = esp_cache_get_alignment(heap_caps, &cache_line_size);
+    return ret == ESP_OK ? cache_line_size : 0;
+}
+
+int make_addr_and_size_align(uint8_t **addr, uint32_t *size)
+{
+    uint8_t *buf = *addr;
+    uint16_t cache_line = get_cache_line_size(buf);
+    uint8_t *buf_alignd = (uint8_t *) ALIGN_DOWN((uint32_t)buf, cache_line);
+    uint32_t len = *size;
+    len += (buf - buf_alignd);
+    uint32_t size_align = ALIGN_UP(len, cache_line);
+
+    *addr = buf_alignd;
+    *size = size_align;
+    return ESP_OK;
+}
+
+static bool IRAM_ATTR dma2d_m2m_suc_eof_event_cb(void *user_data)
+{
+    BaseType_t xHigherPriorityTaskWoken = pdFALSE;
+
+    SemaphoreHandle_t sem = (SemaphoreHandle_t)user_data;
+    xSemaphoreGiveFromISR(sem, &xHigherPriorityTaskWoken);
+    return (xHigherPriorityTaskWoken == pdTRUE);
+}
+
+static void dma2d_link_dscr_init(uint32_t *head, uint32_t *next, void *buf_ptr,
+                                 uint32_t ha, uint32_t va, uint32_t hb, uint32_t vb,
+                                 uint32_t eof, uint32_t en_2d, uint32_t pbyte, uint32_t mod,
+                                 uint32_t bias_x, uint32_t bias_y)
+{
+    dma2d_descriptor_t *dma2d = (dma2d_descriptor_t *)head;
+    memset(dma2d, 0, sizeof(dma2d_descriptor_t));
+    dma2d->owner = DMA2D_DESCRIPTOR_BUFFER_OWNER_DMA;
+    dma2d->suc_eof = eof;
+    dma2d->dma2d_en = en_2d;
+    dma2d->err_eof = 0;
+    dma2d->hb_length = hb;
+    dma2d->vb_size = vb;
+    dma2d->pbyte = pbyte;
+    dma2d->ha_length = ha;
+    dma2d->va_size = va;
+    dma2d->mode = mod;
+    dma2d->y = bias_y;
+    dma2d->x = bias_x;
+    dma2d->buffer = buf_ptr;
+    dma2d->next = (dma2d_descriptor_t *)next;
+}
+
+esp_err_t dma2d_color_convert(dma2d_pic_config_t *src, dma2d_pic_config_t *dest)
+{
+    assert(src->pic_buf);
+    assert(dest->pic_buf);
+
+    int64_t start = esp_timer_get_time();
+    dma2d_m2m_trans_config_t m2m_trans_config = {0};
+
+    if (dma2d_m2m_init() != ESP_OK) {
+        return ESP_FAIL;
+    }
+
+    dma2d_descriptor_t *tx_dsc;
+    dma2d_descriptor_t *rx_dsc;
+
+    // Descriptor and buffer address and size should aligned to 64 bytes (the cacheline size alignment restriction) to be used by CPU
+    dma2d_descriptor_t *tx_link_buffer = (dma2d_descriptor_t *)heap_caps_aligned_calloc(64, 1, 64, MALLOC_CAP_SPIRAM);
+    dma2d_descriptor_t *rx_link_buffer = (dma2d_descriptor_t *)heap_caps_aligned_calloc(64, 1, 64, MALLOC_CAP_SPIRAM);
```

Comment:
The cache line size is configurable. Can be 64 or 128. Can we take the value from sdkconfig.h?